### DISCUSSION TITLE
Defaults with mutable types are "magically" handled

### DISCUSSION THREAD

**Discussion by frankie567** [pydantic/pydantic#2829] [General]
Hi there!

I came across a behavior that surprised me when working with default values for mutable types. I wanted to explain why we shouldn't use a list or a dict as a default for a field but rather `default_factory` to make sure you always get a fresh new instance.

Indeed, in standard Python, when you do something like this:

```py
class A:
    l = [1, 2, 3]

a1 = A()
a1.l.append(4)
print(a1.l)  # [1, 2, 3, 4]

a2 = A()
print(a2.l)  # [1, 2, 3, 4]
```

`a1` and `a2` shares the same reference to the `l` list.

I was 100% sure that I would get the same behavior with Pydantic:

```py
from pydantic import BaseModel

class A(BaseModel):
    l = [1, 2, 3]

a1 = A()
a1.l.append(4)
print(a1.l)  # [1, 2, 3, 4]

a2 = A()
print(a2.l)  # [1, 2, 3]
```

`a2` got a new `l`!

After digging this up a bit, it seems that the method `get_default` is using a deep copy function, which produce this behavior:

https://github.com/samuelcolvin/pydantic/blob/b718e8e62680c2d7a42f1cf946094e1c0fe28c1d/pydantic/fields.py#L371-L372

So here are a few questions that come to my mind:

* Is it an intended behavior? In my opinion, it's very surprising and different from everything I've seen in Python
* If it is, does it mean it's a "valid" usage in Pydantic?

Cheers 😄

**Comment by PrettyWood** [comment]
Hi @frankie567!
Lol when I looked at your code at first sight I thought "hmm he probably should not set a mutable default" 😆 
I understand now that it was actually intended to teach or something...
Anyway back to pydantic!
I think the main reason is that `default_factory` was added in 1.5. Prior to that only `default` was available hence the need to [deepcopy](https://github.com/samuelcolvin/pydantic/blob/v1.4/pydantic/main.py#L839).
I don't know if pydantic should allow this (if it did, it probably should be done in a v2) because even `dataclass` forbids it and raises `ValueError: mutable default <class 'list'> for field l is not allowed: use default_factory`

**Reply by frankie567** [reply]
Exactly, I wanted to show why developers **shouldn't** do that. I was so sure of me that I did the demo live... And it didn't work. Kinda ruined my dramatic effect 😂

I see that we agree that it's not an expected behavior generally-speaking, that reassures me! Thank you for your reply 😀
